docs/markdown/Using Pants/environments.md

```diff
@@ -0,0 +1,221 @@
+)
+```
+
+You'd then give your `remote_environment` target an unassuming name like "default":
+```toml
+[environments.names]
+default = "//:remote_busybox"
+local = "//:local"
+```
+... and use that environment by default with all targets. Users or consumers like CI could then toggle whether remote execution was used by setting `--remote-execution`.
+
+> 🚧 Speculation of remote execution
+> 
+> The `2.15.x` series of Pants does not yet support ["speculating" remote execution](https://github.com/pantsbuild/pants/issues/8353) by racing it against another environment (usually local or docker). While we expect that this will be necessary to make remote execution a viable option for local execution on user's laptops (where network connections are less reliable), it is less critical for CI use-cases.
+
+### Use a `docker_environment` to build the inputs to a `docker_image`
+
+To build a `docker_image` target containing a `pex_binary` which uses native (i.e. compiled) dependencies on a `macOS` machine, you can configure the `pex_binary` to be built in a `docker_environment`.
+
+You'll need a `docker_environment` which uses an image containing the relevant build-time requirements of your PEX. At a minimum, you'll need Python itself:
+```python
+docker_environment(
+  name="python_bullseye",
+  platform="linux_x86_64",
+  image="python:3.9.14-slim-bullseye@sha256-abcd123...",
+  ..
+)
+```
+
+Next, mark your `pex_binary` target with this environment (with the name `python_bullseye`: see "Defining environments" above), and define a `docker_image` target depending on it.
+
+```python
+pex_binary(
+  name="main",
+  environment="python_bullseye",
+)
+
+docker_image(
+    name="docker_image",
+    instructions=[
+        "FROM python:3.9.14-slim-bullseye@sha256-abcd123...",
+        "ENTRYPOINT ["/main"]",
+        "COPY examples/main.pex /main",
+    ],
+)
+```
+
+> 👍 Compatibility of `docker_environment` and `docker_image`
+>
+> Note that the Docker image used in your `docker_environment` does not need to match the base image of the `docker_image` targets that consume them: they only need to be compatible. This is because execution of build steps in a `docker_environment` occurs in an anonymous container, and only the required inputs are provided to the `docker_image` build.
+>
+> This means that your `docker_environment` can include things like compilers or other tools relevant to your build, without needing to manually use multi-stage Docker builds.
+
+### Execute a test in Docker, while natively cross-building it
+
+_TODO: Like https://github.com/pantsbuild/pants/issues/15764, but for tests._
+
+_TODO: Give an example of using `environment=` vs `runtime_environment=` to use docker only for test execution, but not for building of thirdparty dependencies by using PEX to cross-build. This will require exposing options out of PEX to override (?) the target platform, and figuring out how that doesn't break `check` is an open question._
```

Comment:
So: I'm pretty sure that we should avoid stabilizing environments until we have figured out exactly how to implement a cross-build like this. Many other very useful workflows are enabled, but this one in particular feels like it requires a bit more design work.

This is additionally covered by https://github.com/pantsbuild/pants/issues/8384.